dbms/src/TiDB/Raft/RaftLogManager.cpp

```diff
@@ -0,0 +1,184 @@
+    Hints hints;
+    hints.swap(tasks);
+    return hints;
+}
+
+struct RegionGcTask
+{
+    bool skip;
+    UInt64 eager_truncate_index;
+    UInt64 applied_index;
+};
+
+RegionGcTask getRegionTask(
+    const RegionID region_id,
+    const raft_serverpb::RaftApplyState & region_state,
+    const RaftLogEagerGcHint & hint,
+    UInt64 threshold,
+    const LoggerPtr & logger)
+{
+    // LOG_DEBUG(logger, "Load persisted region apply state, region_id={} state={}", region_id, region_state.ShortDebugString());
+    if (region_state.applied_index() <= 1)
+        return RegionGcTask{.skip = true, .eager_truncate_index = 0, .applied_index = 0};
+
+    RegionGcTask task{.skip = false, .eager_truncate_index = 0, .applied_index = 0};
+    task.eager_truncate_index = hint.eager_truncate_index;
+    task.applied_index = region_state.applied_index() - 1;
+    if (task.applied_index <= task.eager_truncate_index || task.applied_index - task.eager_truncate_index <= threshold)
+    {
+        LOG_DEBUG(
+            logger,
+            "Skip eager gc, region_id={} persist_first_index={} persist_applied_index={} hint_eager_truncate_index={} "
+            "hint_applied_index={}",
+            region_id,
+            region_state.truncated_state().index(),
+            region_state.applied_index(),
+            hint.eager_truncate_index,
+            hint.applied_index);
+        task.skip = true;
+    }
+    return task;
+}
+
+RaftLogGcTasksRes executeRaftLogGcTasks(Context & global_ctx, RaftLogEagerGcTasks::Hints && hints)
+{
+    auto write_node_ps = global_ctx.tryGetWriteNodePageStorage();
+    if (!write_node_ps)
+        return {};
+    assert(write_node_ps);
+
+    const auto eager_gc_rows = global_ctx.getTMTContext().getKVStore()->getRaftLogEagerGCRows();
+    LoggerPtr logger = Logger::get();
+
+    Stopwatch watch;
+    RaftLogGcTasksRes eager_truncated_indexes;
+    size_t num_skip = 0;
+    size_t total_num_raft_log_removed = 0;
+    for (const auto & [region_id, hint] : hints)
+    {
+        // Read region apply_state from UniPS
+        auto region_apply_state_key = UniversalPageIdFormat::toRaftApplyStateKeyInKVEngine(region_id);
+        auto page = write_node_ps->read(region_apply_state_key);
+        raft_serverpb::RaftApplyState region_apply_state;
+        if (!region_apply_state.ParseFromString(String(page.data)))
+        {
+            LOG_INFO(logger, "Parse region apply state failed, skip. region_id={}", region_id);
+            continue;
+        }
+
+        UniversalWriteBatch del_batch;
+        const auto region_task = getRegionTask(region_id, region_apply_state, hint, eager_gc_rows, logger);
+        if (region_task.skip)
+        {
+            num_skip += 1;
+            continue;
+        }
+
+        for (UInt64 log_index = region_task.eager_truncate_index; log_index < region_task.applied_index; ++log_index)
+        {
+            del_batch.delPage(UniversalPageIdFormat::toRaftLogKey(region_id, log_index));
```

Comment:
Will the write batch be too large when there are too many entries to be removed? I am afraid it may happens when the `eager_truncate_index` is   not updated, may be after a restart.